// asyncbox/build/lib/asyncbox.js
'use strict';

var _toConsumableArray = require('babel-runtime/helpers/to-consumable-array')['default'];

var _slicedToArray = require('babel-runtime/helpers/sliced-to-array')['default'];

var _Object$defineProperty = require('babel-runtime/core-js/object/define-property')['default'];

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _getIterator = require('babel-runtime/core-js/get-iterator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

_Object$defineProperty(exports, '__esModule', {
  value: true
});

var _q = require('q');

var _q2 = _interopRequireDefault(_q);

var _es6Mapify = require('es6-mapify');

require('source-map-support').install();

function sleep(ms) {
  var nodeSetTimeout;
  return _regeneratorRuntime.async(function sleep$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        nodeSetTimeout = function nodeSetTimeout(ms, cb) {
          return setTimeout(cb, ms);
        };

        context$1$0.next = 3;
        return _regeneratorRuntime.awrap(_q2['default'].nfcall(nodeSetTimeout, ms));

      case 3:
        return context$1$0.abrupt('return', context$1$0.sent);

      case 4:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

function retry(times, fn) {
  for (var _len = arguments.length, args = Array(_len > 2 ? _len - 2 : 0), _key = 2; _key < _len; _key++) {
    args[_key - 2] = arguments[_key];
  }

  var tries, done, res;
  return _regeneratorRuntime.async(function retry$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        tries = 0;
        done = false;
        res = null;

      case 3:
        if (!(!done && tries < times)) {
          context$1$0.next = 18;
          break;
        }

        tries++;
        context$1$0.prev = 5;
        context$1$0.next = 8;
        return _regeneratorRuntime.awrap(fn.apply(undefined, args));

      case 8:
        res = context$1$0.sent;

        done = true;
        context$1$0.next = 16;
        break;

      case 12:
        context$1$0.prev = 12;
        context$1$0.t0 = context$1$0['catch'](5);

        if (!(tries >= times)) {
          context$1$0.next = 16;
          break;
        }

        throw context$1$0.t0;

      case 16:
        context$1$0.next = 3;
        break;

      case 18:
        return context$1$0.abrupt('return', res);

      case 19:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this, [[5, 12]]);
}

function retryInterval(times, sleepMs, fn) {
  for (var _len2 = arguments.length, args = Array(_len2 > 3 ? _len2 - 3 : 0), _key2 = 3; _key2 < _len2; _key2++) {
    args[_key2 - 3] = arguments[_key2];
  }

  var wrapped;
  return _regeneratorRuntime.async(function retryInterval$(context$1$0) {
    var _this = this;

    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        wrapped = function wrapped() {
          var res;
          return _regeneratorRuntime.async(function wrapped$(context$2$0) {
            while (1) switch (context$2$0.prev = context$2$0.next) {
              case 0:
                res = undefined;
                context$2$0.prev = 1;
                context$2$0.next = 4;
                return _regeneratorRuntime.awrap(fn.apply(undefined, args));

              case 4:
                res = context$2$0.sent;
                context$2$0.next = 12;
                break;

              case 7:
                context$2$0.prev = 7;
                context$2$0.t0 = context$2$0['catch'](1);
                context$2$0.next = 11;
                return _regeneratorRuntime.awrap(sleep(sleepMs));

              case 11:
                throw context$2$0.t0;

              case 12:
                return context$2$0.abrupt('return', res);

              case 13:
              case 'end':
                return context$2$0.stop();
            }
          }, null, _this, [[1, 7]]);
        };

        return context$1$0.abrupt('return', retry(times, wrapped));

      case 2:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

function parallel(promises) {
  var done, results, _iteratorNormalCompletion, _didIteratorError, _iteratorError, _iterator, _step, p;

  return _regeneratorRuntime.async(function parallel$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        if (promises instanceof Array) {
          context$1$0.next = 2;
          break;
        }

        throw new Error('Must parallelize a list of promises');

      case 2:
        if (!(promises.length < 1)) {
          context$1$0.next = 4;
          break;
        }

        throw new Error('List of promises was empty');

      case 4:
        done = _q2['default'].defer();
        results = [];
        _iteratorNormalCompletion = true;
        _didIteratorError = false;
        _iteratorError = undefined;
        context$1$0.prev = 9;
        _iterator = _getIterator(promises);

      case 11:
        if (_iteratorNormalCompletion = (_step = _iterator.next()).done) {
          context$1$0.next = 19;
          break;
        }

        p = _step.value;

        if (!(!p.then || !p['catch'])) {
          context$1$0.next = 15;
          break;
        }

        throw new Error('Tried to parallelize a non-promise');

      case 15:
        p.then(function (res) {
          results.push(res);
          if (results.length === promises.length) {
            done.resolve(results);
          }
        }, function (err) {
          done.reject(err);
        });

      case 16:
        _iteratorNormalCompletion = true;
        context$1$0.next = 11;
        break;

      case 19:
        context$1$0.next = 25;
        break;

      case 21:
        context$1$0.prev = 21;
        context$1$0.t0 = context$1$0['catch'](9);
        _didIteratorError = true;
        _iteratorError = context$1$0.t0;

      case 25:
        context$1$0.prev = 25;
        context$1$0.prev = 26;

        if (!_iteratorNormalCompletion && _iterator['return']) {
          _iterator['return']();
        }

      case 28:
        context$1$0.prev = 28;

        if (!_didIteratorError) {
          context$1$0.next = 31;
          break;
        }

        throw _iteratorError;

      case 31:
        return context$1$0.finish(28);

      case 32:
        return context$1$0.finish(25);

      case 33:
        return context$1$0.abrupt('return', done.promise);

      case 34:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this, [[9, 21, 25, 33], [26,, 28, 32]]);
}

function nodeify(promisey, cb) {
  promisey.then(function (res) {
    cb(null, res);
  }, cb);
}

function nodeifyAll(promiseyMap) {
  var cbMap = {};
  var _iteratorNormalCompletion2 = true;
  var _didIteratorError2 = false;
  var _iteratorError2 = undefined;

  try {
    var _loop = function () {
      var _step2$value = _slicedToArray(_step2.value, 2);

      var name = _step2$value[0];
      var fn = _step2$value[1];

      /*jshint -W083 */
      cbMap[name] = function () {
        for (var _len3 = arguments.length, args = Array(_len3), _key3 = 0; _key3 < _len3; _key3++) {
          args[_key3] = arguments[_key3];
        }

        var _cb = args.slice(-1)[0];
        args = args.slice(0, -1);
        nodeify(fn.apply(undefined, _toConsumableArray(args)), _cb);
      };
    };

    for (var _iterator2 = _getIterator((0, _es6Mapify.mapify)(promiseyMap)), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
      _loop();
    }
  } catch (err) {
    _didIteratorError2 = true;
    _iteratorError2 = err;
  } finally {
    try {
      if (!_iteratorNormalCompletion2 && _iterator2['return']) {
        _iterator2['return']();
      }
    } finally {
      if (_didIteratorError2) {
        throw _iteratorError2;
      }
    }
  }

  return cbMap;
}

function asyncify(fn) {
  for (var _len4 = arguments.length, args = Array(_len4 > 1 ? _len4 - 1 : 0), _key4 = 1; _key4 < _len4; _key4++) {
    args[_key4 - 1] = arguments[_key4];
  }

  fn.call.apply(fn, [null].concat(args)).then(function () {}, function (err) {
    console.error(err.stack);
  });
}

exports.sleep = sleep;
exports.retry = retry;
exports.nodeify = nodeify;
exports.nodeifyAll = nodeifyAll;
exports.retryInterval = retryInterval;
exports.asyncify = asyncify;
exports.parallel = parallel;

/*jshint -W083 */
